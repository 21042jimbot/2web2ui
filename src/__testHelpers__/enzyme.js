export const findButton = (wrapper, buttonLabel) =>
  wrapper.findWhere(
    (node) => node.type() === 'button' && new RegExp(buttonLabel).test(node.text())
  );
