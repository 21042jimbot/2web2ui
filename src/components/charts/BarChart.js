/* eslint-disable no-unused-vars */
import React from 'react';
import {
  Bar,
  CartesianGrid,
  ComposedChart,
  Rectangle,
  ReferenceLine,
  ResponsiveContainer,
  Tooltip,
  XAxis,
  YAxis
} from 'recharts';
import TooltipWrapper from 'src/pages/signals/components/charts/tooltip/Tooltip';
import 'src/pages/signals/components/charts/barchart/BarChart.scss';

import _ from 'lodash';
import PropTypes from 'prop-types';

export const BarChart = (props) => {
  const {
    cartesianGridProps,
    disableHover,
    gap,
    height,
    hovered,
    isLink,
    margin,
    onClick,
    onMouseOver,
    timeSeries,
    tooltipContent,
    tooltipWidth,
    width,
    xAxisRefLines,
    yAxisRefLines,
    xKey,
    xAxisProps,
    yDomain,
    yAxisProps,
    children
  } = props;

  return (
    <ResponsiveContainer height={height} width={width} className='SignalsBarChart'>
      <ComposedChart
        barCategoryGap={gap}
        data={timeSeries}
        margin={margin}
        cursor={(isLink) ? 'pointer' : 'default'}
      >
        <CartesianGrid
          vertical={false}
          stroke='#e1e1e6'
          shapeRendering='crispEdges'
          {...cartesianGridProps}
        />
        <YAxis
          axisLine={false}
          tickLine={false}
          width={30}
          minTickGap={2}
          domain={yDomain}
          {...yAxisProps}
        />
        <XAxis
          axisLine={false}
          tickLine={false}
          dataKey={xKey}
          type='category'
          padding={{ left: 12, right: 12 }}
          shapeRendering='crispEdges'
          {...xAxisProps}
        />
        {!disableHover && hovered && (
          <Tooltip
            offset={25}
            cursor={false}
            isAnimationActive={false}
            content={<TooltipWrapper children={tooltipContent} />}
            width={tooltipWidth}
            position={{ x: 0, y: 0 }}
          />
        )}
        {xAxisRefLines.length && (
          _.map(xAxisRefLines, (xAxisRefLine, index) =>
            <ReferenceLine
              key={`x-${index}`}
              style={{ pointerEvents: 'none' }}
              x={xAxisRefLine.x}
              shapeRendering='crispEdges'
              stroke={xAxisRefLine.stroke}
              strokeWidth={xAxisRefLine.strokeWidth}
              label={xAxisRefLine.label}
            />
          )
        )}
        {yAxisRefLines.length && (
          _.map(yAxisRefLines, (yAxisRefLine, index) =>
            <ReferenceLine
              key={`y-${index}`}
              style={{ pointerEvents: 'none' }}
              y={yAxisRefLine.y}
              shapeRendering='crispEdges'
              stroke={yAxisRefLine.stroke}
              strokeWidth={yAxisRefLine.strokeWidth}
            />
          )
        )}
        <Bar
          cursor='pointer'
          dataKey='noKey'
          stackId='stack'
          isAnimationActive={false}
          onClick={onClick}
          onMouseOver={onMouseOver}
          shape={
            ({ payload, background, ...rest }) => (
              <Rectangle {...rest} {...background} opacity={0} />
            )
          }
        />
        {children}
      </ComposedChart>
    </ResponsiveContainer>
  );
};
BarChart.propTypes = {
  activeFill: PropTypes.string,
  fill: PropTypes.string,
  gap: PropTypes.number,
  onClick: PropTypes.func,
  shouldHighlightSelected: PropTypes.bool,
  tooltipContent: PropTypes.func,
  tooltipWidth: PropTypes.string,
  yKeys: PropTypes.arrayOf(PropTypes.object)
};

BarChart.defaultProps = {
  activeFill: '#22838A',
  fill: '#B3ECEF',
  gap: 1,
  height: 250,
  margin: { top: 12, left: 18, right: 0, bottom: 0 },
  shouldHighlightSelected: true,
  width: '99%',
  xAxisRefLines: [],
  yAxisRefLines: [],
  xKey: 'date',
  yKey: 'value',
  yRange: ['auto', 'auto']
};
export default BarChart;
