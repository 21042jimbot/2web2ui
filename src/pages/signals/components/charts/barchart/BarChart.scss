.SignalsBarChart {
  .recharts-bar-rectangle path {
    transition: 0.15s;
  }

  .recharts-text.recharts-cartesian-axis-tick-value {
    font-size: 11px;
  }
}
