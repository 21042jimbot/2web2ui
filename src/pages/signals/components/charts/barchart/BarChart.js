import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { ResponsiveContainer, BarChart as RechartsBarchart, Bar, Tooltip, XAxis, YAxis, CartesianGrid, Rectangle } from 'recharts';
import TooltipWrapper from '../tooltip/Tooltip';
import './BarChart.scss';

/**
 * @example
 * <BarChart timeSeries={data}
 *   yDomain={[0,100]}
 *   yKeys={[
 *    { key: 'bar', fill: '#B157CE' },
 *    { key: 'foo', fill: '#28C0C4' },
 *   ]}
 *   gap={0.5}
 *   onClick={this.handleBarClick}
 *   selected={this.state.selected}
 *   xAxisProps={{ interval: 88, tickFormatter: (tick) => moment(tick).format('M/D') }}
 * />
 */
class BarChart extends Component {
  renderBar = ({ key, fill }) => (

    <Bar
      cursor='pointer'
      stackId='stack'
      key={key}
      dataKey={key}
      onClick={this.props.onClick}
      fill={fill}
      isAnimationActive={false}
      minPointSize={1}
    />

  )

  renderBars = () => {
    const { yKeys, yKey, fill } = this.props;

    if (yKeys) {
      return yKeys.map(this.renderBar);
    }

    return this.renderBar({ key: yKey, fill });
  }

  renderBackgrounds = () => {
    const { xKey, selected, onClick } = this.props;

    return (
      <Bar
        cursor='pointer'
        dataKey='noKey'
        stackId='stack'
        fill='#5DCFF5'
        isAnimationActive={false}
        onClick={onClick}
        shape={
          ({ payload, background, ...rest }) => (
            <Rectangle {...rest} {...background} opacity={payload[xKey] === selected ? 0.4 : 0} />
          )
        }
      />
    );
  }

  render() {
    const { gap, height, disableHover, margin, timeSeries, tooltipContent, tooltipWidth, width, xKey, xAxisProps, yDomain, yAxisProps } = this.props;

    return (
      <ResponsiveContainer height={height} width={width} className='SignalsBarChart'>
        <RechartsBarchart
          barCategoryGap={gap}
          data={timeSeries}
          margin={margin}
        >
          {this.renderBackgrounds()}
          <CartesianGrid
            vertical={false}
            stroke='#e1e1e6'
            shapeRendering='crispEdges'
          />
          <YAxis
            axisLine={false}
            tickLine={false}
            width={30}
            minTickGap={2}
            domain={yDomain}
            {...yAxisProps}
          />
          <XAxis
            axisLine={false}
            tickLine={false}
            dataKey={xKey}
            type='category'
            padding={{ left: 12, right: 12 }}
            shapeRendering='crispEdges'
            {...xAxisProps}
          />
          {!disableHover && (
            <Tooltip
              offset={25}
              cursor={false}
              isAnimationActive={false}
              content={<TooltipWrapper children={tooltipContent} />}
              width={tooltipWidth}
            />
          )}
          {this.renderBars()}
        </RechartsBarchart>
      </ResponsiveContainer>
    );
  }
}

BarChart.propTypes = {
  fill: PropTypes.string,
  gap: PropTypes.number,
  onClick: PropTypes.func,
  tooltipContent: PropTypes.func,
  tooltipWidth: PropTypes.string,
  yKeys: PropTypes.arrayOf(PropTypes.object)
};

BarChart.defaultProps = {
  fill: '#B157CE',
  gap: 1,
  height: 250,
  width: '99%',
  margin: { top: 12, left: 18, right: 0, bottom: 5 },
  xKey: 'date',
  yKey: 'value',
  yRange: ['auto', 'auto']
};

export default BarChart;
