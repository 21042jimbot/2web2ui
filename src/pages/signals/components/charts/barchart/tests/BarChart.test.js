import { shallow } from 'enzyme';
import React from 'react';
import BarChart from '../BarChart';

describe('BarChart Component', () => {
  let wrapper;
  let props;
  const normal = [{ value: 2, date: '2011-01-01' }];
  const stacked = [
    { ab: 2, cd: 3, date: '2011-01-01' },
    { ab: 5, cd: 8, date: '2011-01-02' }
  ];
  const yKeys = [
    { key: 'ab', fill: 'blue' },
    { key: 'cd', fill: 'red' }
  ];

  beforeEach(() => {
    props = {
      timeSeries: normal,
      xKey: 'date',
      yRange: [0,100],
      height: 50,
      width: 100,
      onClick: jest.fn(),
      tooltipContent: jest.fn(),
      fill: '#fill',
      activeFill: '#activeFill'
    };
    wrapper = shallow(<BarChart {...props}/>);
  });

  it('renders a normal bar chart correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders a normal bar chart with correct fill for selected/hovered threshold events', () => {
    wrapper.setProps({ selected: '2011-01-01', yKey: 'health_score', timeSeries: [{ health_score: 75, ranking: 'warning', date: '2011-01-01' }]});
    const payload = { fill: '#fill', health_score: 75, ranking: 'warning', date: '2011-01-01' };
    expect(wrapper.find('Bar').at(0).props().shape(payload).props).toMatchSnapshot();
  });

  it('renders a normal bar chart with correct fill for non-selected/hovered threshold events', () => {
    wrapper.setProps({ selected: undefined, yKey: 'health_score', timeSeries: [{ health_score: 75, ranking: 'warning', date: '2011-01-01' }]});
    const payload = { fill: '#fill', health_score: 75, ranking: 'warning', date: '2011-01-01' };
    expect(wrapper.find('Bar').at(0).props().shape(payload).props).toMatchSnapshot();
  });

  it('renders a normal bar chart with correct fill for selected/hovered non-threshold events', () => {
    wrapper.setProps({ selected: '2011-01-01', yKey: 'injections', timeSeries: [{ injections: 75, date: '2011-01-01' }]});
    const payload = { fill: '#fill', injections: 75, ranking: 'warning', date: '2011-01-01' };
    expect(wrapper.find('Bar').at(0).props().shape(payload).props).toMatchSnapshot();
  });

  it('renders a normal bar chart with correct fill for non-selected/hovered non-threshold events', () => {
    wrapper.setProps({ selected: undefined, yKey: 'injections', timeSeries: [{ injections: 75, date: '2011-01-01' }]});
    const payload = { fill: '#fill', injections: 75, ranking: 'warning', date: '2011-01-01' };
    expect(wrapper.find('Bar').at(0).props().shape(payload).props.fill).toEqual('#fill');
  });

  it('renders a stacked bar chart correctly', () => {
    wrapper.setProps({ timeSeries: stacked, yKeys });
    const bars = wrapper.find('Bar').slice(1);
    expect(bars).toMatchSnapshot();
  });

  it('renders background bars with no opacity if unselected', () => {
    const payload = { payload: { date: '2011-01-01' }, test: 'test' };
    expect(wrapper.find('Bar').at(0).props().shape(payload)).toMatchSnapshot();
  });

  it('renders background bars with opacity if selected', () => {
    wrapper.setProps({ selected: '2011-01-01' });
    const payload = { payload: { date: '2011-01-01' }, test: 'test' };
    expect(wrapper.find('Bar').at(0).props().shape(payload)).toMatchSnapshot();
  });

  it('should handle click on all bars', () => {
    wrapper.setProps({ timeSeries: stacked, yKeys });
    wrapper.find('Bar').forEach((n) => n.simulate('click'));
    expect(props.onClick).toHaveBeenCalledTimes(3);
  });

  it('should display x reference lines', () => {
    wrapper.setProps({ xAxisRefLines: [{ x: '2011-01-01', stroke: 'green', strokeWidth: 2 }, { x: '2011-01-03', stroke: 'red', strokeWidth: 2 }]});
    expect(wrapper).toMatchSnapshot();
  });

  it('should display y reference lines', () => {
    wrapper.setProps({ yAxisRefLines: [{ y: 80, stroke: 'green', strokeWidth: 2 }, { y: 55, stroke: 'red', strokeWidth: 2 }]});
    expect(wrapper).toMatchSnapshot();
  });
});
