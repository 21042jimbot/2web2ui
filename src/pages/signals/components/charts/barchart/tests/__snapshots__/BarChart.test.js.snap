// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BarChart Component renders a normal bar chart correctly 1`] = `
<ResponsiveContainer
  className="SignalsBarChart"
  debounce={0}
  height={50}
  width={100}
>
  <BarChart
    barCategoryGap={1}
    barGap={4}
    data={
      Array [
        Object {
          "date": "2011-01-01",
          "value": 2,
        },
      ]
    }
    layout="horizontal"
    margin={
      Object {
        "bottom": 5,
        "left": 18,
        "right": 0,
        "top": 5,
      }
    }
    reverseStackOrder={false}
    stackOffset="none"
  >
    <Bar
      animationBegin={0}
      animationDuration={400}
      animationEasing="ease"
      cursor="pointer"
      data={Array []}
      dataKey="noKey"
      fill="#5DCFF5"
      hide={false}
      isAnimationActive={false}
      layout="vertical"
      legendType="rect"
      minPointSize={0}
      onAnimationEnd={[Function]}
      onAnimationStart={[Function]}
      onClick={[MockFunction]}
      shape={[Function]}
      stackId="stack"
      xAxisId={0}
      yAxisId={0}
    />
    <CartesianGrid
      fill="none"
      horizontal={true}
      horizontalFill={Array []}
      horizontalPoints={Array []}
      shapeRendering="crispEdges"
      stroke="#e1e1e6"
      vertical={false}
      verticalFill={Array []}
      verticalPoints={Array []}
    />
    <YAxis
      allowDataOverflow={false}
      allowDecimals={true}
      allowDuplicatedCategory={true}
      axisLine={false}
      domain={
        Array [
          0,
          "auto",
        ]
      }
      height={0}
      hide={false}
      minTickGap={2}
      mirror={false}
      orientation="left"
      padding={
        Object {
          "bottom": 0,
          "top": 0,
        }
      }
      reversed={false}
      scale="auto"
      tickCount={5}
      tickLine={false}
      type="number"
      width={30}
      yAxisId={0}
    />
    <XAxis
      allowDataOverflow={false}
      allowDecimals={true}
      allowDuplicatedCategory={true}
      axisLine={false}
      dataKey="date"
      domain={
        Array [
          0,
          "auto",
        ]
      }
      height={30}
      hide={false}
      mirror={false}
      orientation="bottom"
      padding={
        Object {
          "left": 12,
          "right": 12,
        }
      }
      reversed={false}
      scale="auto"
      shapeRendering="crispEdges"
      tickCount={5}
      tickLine={false}
      type="category"
      width={0}
      xAxisId={0}
    />
    <Tooltip
      active={false}
      animationDuration={400}
      animationEasing="ease"
      content={
        <Tooltip>
          [MockFunction]
        </Tooltip>
      }
      contentStyle={Object {}}
      coordinate={
        Object {
          "x": 0,
          "y": 0,
        }
      }
      cursor={false}
      cursorStyle={Object {}}
      filterNull={true}
      isAnimationActive={false}
      itemSorter={[Function]}
      itemStyle={Object {}}
      labelStyle={Object {}}
      offset={25}
      separator=" : "
      useTranslate3d={false}
      viewBox={
        Object {
          "x1": 0,
          "x2": 0,
          "y1": 0,
          "y2": 0,
        }
      }
      wrapperStyle={Object {}}
    />
    <Bar
      animationBegin={0}
      animationDuration={400}
      animationEasing="ease"
      cursor="pointer"
      data={Array []}
      dataKey="value"
      fill="#fill"
      hide={false}
      isAnimationActive={false}
      key="value"
      layout="vertical"
      legendType="rect"
      minPointSize={1}
      onAnimationEnd={[Function]}
      onAnimationStart={[Function]}
      onClick={[MockFunction]}
      stackId="stack"
      xAxisId={0}
      yAxisId={0}
    />
  </BarChart>
</ResponsiveContainer>
`;

exports[`BarChart Component renders a stacked bar chart correctly 1`] = `
Array [
  <Bar
    animationBegin={0}
    animationDuration={400}
    animationEasing="ease"
    cursor="pointer"
    data={Array []}
    dataKey="ab"
    fill="blue"
    hide={false}
    isAnimationActive={false}
    key="ab"
    layout="vertical"
    legendType="rect"
    minPointSize={1}
    onAnimationEnd={[Function]}
    onAnimationStart={[Function]}
    onClick={[MockFunction]}
    stackId="stack"
    xAxisId={0}
    yAxisId={0}
  />,
  <Bar
    animationBegin={0}
    animationDuration={400}
    animationEasing="ease"
    cursor="pointer"
    data={Array []}
    dataKey="cd"
    fill="red"
    hide={false}
    isAnimationActive={false}
    key="cd"
    layout="vertical"
    legendType="rect"
    minPointSize={1}
    onAnimationEnd={[Function]}
    onAnimationStart={[Function]}
    onClick={[MockFunction]}
    stackId="stack"
    xAxisId={0}
    yAxisId={0}
  />,
]
`;

exports[`BarChart Component renders background bars with no opacity if unselected 1`] = `
<Rectangle
  animationBegin={0}
  animationDuration={1500}
  animationEasing="ease"
  height={0}
  isAnimationActive={false}
  isUpdateAnimationActive={false}
  opacity={0}
  radius={0}
  test="test"
  width={0}
  x={0}
  y={0}
/>
`;

exports[`BarChart Component renders background bars with opacity if selected 1`] = `
<Rectangle
  animationBegin={0}
  animationDuration={1500}
  animationEasing="ease"
  height={0}
  isAnimationActive={false}
  isUpdateAnimationActive={false}
  opacity={0.4}
  radius={0}
  test="test"
  width={0}
  x={0}
  y={0}
/>
`;
